lib/tasks/dev.rake

```diff
@@ -129,10 +131,25 @@ namespace :dev do
               )
               enrollment.save!
 
-              UspsInPersonProofing::EnrollmentHelper.schedule_in_person_enrollment(
-                user,
-                pii,
-              )
+              success = false
+              num_attempts = 0
+              max_attempts = (ENV['MAX_NUM_ATTEMPTS'] || 3).to_i
+              max_attempts.freeze
+              until success || num_attempts >= max_attempts
+                num_attempts += 1
+                begin
+                  UspsInPersonProofing::EnrollmentHelper.schedule_in_person_enrollment(
+                    user,
+                    pii,
+                  )
+                rescue StandardError => e
+                  Rails.logger.error 'Exception raised while enrolling user: ' + e.message
+                  raise e if num_attempts == max_attempts
+                else
+                  success = true
+                end
+                sleep(usps_request_delay_ms / 1000.0) if usps_request_delay_ms
```

Comment:
I recommend accounting for the computing time as part of the delay instead of always delaying a full second.

For example:
```ruby
last_request_completed_time = nil

# in loop...
begin
  sleep([usps_request_delay_ms  - (Time.now.to_f - last_request_completed_time), 0].max) unless last_request_completed_time.nil?
  UspsInPersonProofing::EnrollmentHelper.schedule_in_person_enrollment(user, pii)
ensure
  last_request_completed_time = Time.now.to_f
end
```

Reply:
At first I liked this idea but I'm not sure I do anymore. Can you suggests how I should update the tests to account for `sleep()` getting called with a variable value per this suggestion?

My arguments against it are that it makes the code more complicated, it will make the tests more complicated, and all it's really doing is smoothing out delays that will otherwise have some random jitter in them. I actually liked the article you linked a couple months back about the value of jitter in delays -- why is it bad here?

Reply:
You can adjust the existing test for this by asserting that the sleep time is less than or equal to the delay.

If you want to more specifically test this behavior, then you could can stub `Time.now` and verify that `sleep` gets called with the expected difference in times.

Regarding jitter - I don't think that the delays implicitly added by the network call & compute time serve the same purpose as jitter. If we want to add jitter, then we could add a configurable quantity added to the delay that varies between something like `-50%` and `+50%` of the total delay.

The original delay as I understand it is meant to control network traffic to the USPS API, which isn't served by unnecessarily taking into account CPU load of the worker. Usually it's better to serve the call at the fastest acceptable pace so concurrent tasks don't stack up.